central/auth/service/service_impl_test.go

```diff
@@ -126,3 +181,428 @@ func (s *authServiceAccessControlTestSuite) TestAuthServiceResponse() {
 		})
 	}
 }
+
+func (s *authServiceAccessControlTestSuite) TestValidateAuthMachineToMachineConfig() {
+	testCases := []struct {
+		config      *v1.AuthMachineToMachineConfig
+		skipIDCheck bool
+		err         error
+	}{
+		{
+			config: nil,
+			err:    errox.InvalidArgs,
+		},
+		{
+			config: &v1.AuthMachineToMachineConfig{},
+			err:    errox.InvalidArgs,
+		},
+		{
+			config: &v1.AuthMachineToMachineConfig{
+				Id: "some-id",
+			},
+			err: errox.InvalidArgs,
+		},
+		{
+			config: &v1.AuthMachineToMachineConfig{
+				Id:                      "some-id",
+				TokenExpirationDuration: "1",
+			},
+			err: errox.InvalidArgs,
```

Comment:
I agree with the sentiment, but disagree with your proposed approach.

In a typed language relying on error message strings to be constructed a certain way seems just _meh_.

Instead, we should create a custom error type and wrap it properly, this way we can then check for the customer error type specifically here instead of relying on regular expressions or rather the error message.

The new wrapped error will look like this: `invalid arguments: invalid token expiration duration given: time: missing unit in duration "1"`.

Reply:
If I control both the error text and the test code, I find it's ok to use regex to check the resulting error text in test.

I don't mind the way you approach this. I just want to mention that error messages are for developers to understand what went wrong and error types are for developers to distinguish and handle error situations in the calling code.
If there is no calling code that needs to distinguish and handle error situations, I would not add error types because I find that to be early over-engineering.

Another advantage of checking error message in tests is that it makes me look at the error message and read it. This makes me review whether error messages would make sense for developers without prior knowledge. You seemed to look at that too, at least in that example.

Shared my opinion. Feel free to resolve this thread.